File: pages/addpatient.py
from PyQt5 import QtWidgets, QtCore, QtGui
import sys
import datetime
import requests
import json

class AddPatientDialog(QtWidgets.QDialog):
    """Dialog for adding a new patient to the database."""
    def __init__(self, parent=None):
        super().__init__(parent)
        self.setWindowTitle("Add New Patient")
        self.setMinimumSize(600, 650)  # Increased height for additional fields
        self.setup_ui()
        self.connect_signals()
        
    def setup_ui(self):
        # Set window properties
        self.setStyleSheet("""
            QDialog {
                background-color: #f0f7ff;
                border: 1px solid #3498db;
                border-radius: 10px;
            }
            QLabel {
                font-size: 14px;
                color: #2c3e50;
            }
            QLineEdit, QDateEdit, QTextEdit, QDateTimeEdit, QComboBox {
                border: 2px solid #bdc3c7;
                border-radius: 6px;
                padding: 8px;
                font-size: 14px;
                background-color: white;
            }
            QLineEdit:focus, QDateEdit:focus, QTextEdit:focus, QDateTimeEdit:focus, QComboBox:focus {
                border: 2px solid #3498db;
            }
            QPushButton {
                border-radius: 6px;
                padding: 10px 20px;
                font-size: 14px;
                font-weight: bold;
            }
            QGroupBox {
                font-weight: bold;
                border: 1px solid #bdc3c7;
                border-radius: 6px;
                margin-top: 12px;
            }
            QGroupBox::title {
                subcontrol-origin: margin;
                left: 10px;
                padding: 0 5px 0 5px;
            }
        """)
        
        # Main layout
        main_layout = QtWidgets.QVBoxLayout(self)
        main_layout.setContentsMargins(25, 25, 25, 25)
        main_layout.setSpacing(15)
        
        # Title with icon
        title_layout = QtWidgets.QHBoxLayout()
        icon_label = QtWidgets.QLabel("👤+")
        icon_label.setStyleSheet("font-size: 24px;")
        title_label = QtWidgets.QLabel("Add New Patient")
        title_label.setStyleSheet("font-size: 20px; font-weight: bold; color: #2980b9;")
        
        title_layout.addWidget(icon_label)
        title_layout.addWidget(title_label)
        title_layout.addStretch()
        main_layout.addLayout(title_layout)
        
        # Separator line
        line = QtWidgets.QFrame()
        line.setFrameShape(QtWidgets.QFrame.HLine)
        line.setFrameShadow(QtWidgets.QFrame.Sunken)
        line.setStyleSheet("background-color: #3498db;")
        main_layout.addWidget(line)
        
        # Create a form layout for patient details
        form_layout = QtWidgets.QFormLayout()
        form_layout.setLabelAlignment(QtCore.Qt.AlignRight | QtCore.Qt.AlignVCenter)
        form_layout.setFormAlignment(QtCore.Qt.AlignLeft | QtCore.Qt.AlignVCenter)
        form_layout.setSpacing(15)
        
        # Patient Name field
        name_label = QtWidgets.QLabel("Name:")
        name_label.setStyleSheet("font-weight: bold;")
        self.name_input = QtWidgets.QLineEdit()
        self.name_input.setPlaceholderText("Enter patient's name")
        form_layout.addRow(name_label, self.name_input)
        
        # Birth Date field
        birth_date_label = QtWidgets.QLabel("Birth Date:")
        birth_date_label.setStyleSheet("font-weight: bold;")
        self.birth_date_input = QtWidgets.QDateEdit(calendarPopup=True)
        self.birth_date_input.setDisplayFormat("yyyy-MM-dd")
        self.birth_date_input.setDate(QtCore.QDate.currentDate().addYears(-30))  # Default to 30 years ago
        self.birth_date_input.setMaximumDate(QtCore.QDate.currentDate())
        form_layout.addRow(birth_date_label, self.birth_date_input)
        
        # Admission Date field
        admission_date_label = QtWidgets.QLabel("Admission Date:")
        admission_date_label.setStyleSheet("font-weight: bold;")
        self.admission_date_input = QtWidgets.QDateEdit(calendarPopup=True)
        self.admission_date_input.setDisplayFormat("yyyy-MM-dd")
        self.admission_date_input.setDate(QtCore.QDate.currentDate())  # Default to today
        form_layout.addRow(admission_date_label, self.admission_date_input)
        
        # Status field
        status_label = QtWidgets.QLabel("Status:")
        status_label.setStyleSheet("font-weight: bold;")
        self.status_input = QtWidgets.QComboBox()
        self.status_input.addItems(['STABLE', 'CRITICAL', 'URGENT', 'GOOD', 'RECOVERING'])
        form_layout.addRow(status_label, self.status_input)
        
        # Notes field
        notes_label = QtWidgets.QLabel("Notes:")
        notes_label.setStyleSheet("font-weight: bold;")
        self.notes_input = QtWidgets.QTextEdit()
        self.notes_input.setPlaceholderText("Enter any notes about the patient")
        self.notes_input.setMaximumHeight(100)
        form_layout.addRow(notes_label, self.notes_input)
        
        # Add form layout to main layout
        main_layout.addLayout(form_layout)
        
        # Disease group box
        disease_group = QtWidgets.QGroupBox("Disease")
        disease_layout = QtWidgets.QVBoxLayout(disease_group)
        
        self.disease_input = QtWidgets.QLineEdit()
        self.disease_input.setPlaceholderText("Enter patient's disease")
        disease_layout.addWidget(self.disease_input)
        
        main_layout.addWidget(disease_group)
        
        # Symptoms group box
        symptoms_group = QtWidgets.QGroupBox("Symptoms")
        symptoms_layout = QtWidgets.QVBoxLayout(symptoms_group)
        
        # Symptoms input layout
        self.symptoms_list = QtWidgets.QListWidget()
        self.symptoms_list.setStyleSheet("""
            QListWidget {
                border: 2px solid #bdc3c7;
                border-radius: 6px;
                background-color: white;
            }
            QListWidget::item {
                padding: 5px;
                border-bottom: 1px solid #ecf0f1;
            }
            QListWidget::item:selected {
                background-color: #3498db;
                color: white;
            }
        """)
        
        symptoms_input_layout = QtWidgets.QHBoxLayout()
        self.symptom_input = QtWidgets.QLineEdit()
        self.symptom_input.setPlaceholderText("Enter a symptom")
        
        self.severity_input = QtWidgets.QComboBox()
        self.severity_input.addItems(['Mild', 'Moderate', 'Severe', 'Critical'])
        
        self.add_symptom_btn = QtWidgets.QPushButton("Add")
        self.add_symptom_btn.setStyleSheet("""
            QPushButton {
                background-color: #2ecc71;
                color: white;
            }
            QPushButton:hover {
                background-color: #27ae60;
            }
        """)
        
        self.remove_symptom_btn = QtWidgets.QPushButton("Remove")
        self.remove_symptom_btn.setStyleSheet("""
            QPushButton {
                background-color: #e74c3c;
                color: white;
            }
            QPushButton:hover {
                background-color: #c0392b;
            }
        """)
        
        symptoms_input_layout.addWidget(self.symptom_input)
        symptoms_input_layout.addWidget(self.severity_input)
        symptoms_input_layout.addWidget(self.add_symptom_btn)
        symptoms_input_layout.addWidget(self.remove_symptom_btn)
        
        symptoms_layout.addLayout(symptoms_input_layout)
        symptoms_layout.addWidget(self.symptoms_list)
        
        main_layout.addWidget(symptoms_group)
        
        # Button layout
        button_layout = QtWidgets.QHBoxLayout()
        button_layout.setSpacing(15)
        
        self.cancel_button = QtWidgets.QPushButton("Cancel")
        self.cancel_button.setStyleSheet("""
            QPushButton {
                background-color: #e74c3c;
                color: white;
                min-height: 35px;
                min-width: 100px;
            }
            QPushButton:hover {
                background-color: #c0392b;
            }
        """)
        self.cancel_button.setIcon(self.style().standardIcon(QtWidgets.QStyle.SP_DialogCancelButton))
        
        self.save_button = QtWidgets.QPushButton("Save Patient")
        self.save_button.setStyleSheet("""
            QPushButton {
                background-color: #3498db;
                color: white;
                min-height: 35px;
                min-width: 150px;
            }
            QPushButton:hover {
                background-color: #2980b9;
            }
        """)
        self.save_button.setIcon(self.style().standardIcon(QtWidgets.QStyle.SP_DialogSaveButton))
        self.save_button.setDefault(True)
        
        button_layout.addStretch()
        button_layout.addWidget(self.cancel_button)
        button_layout.addWidget(self.save_button)
        
        main_layout.addLayout(button_layout)
    
    def connect_signals(self):
        """Connect UI element signals to slots."""
        self.cancel_button.clicked.connect(self.reject)
        self.save_button.clicked.connect(self.save_patient)
        self.add_symptom_btn.clicked.connect(self.add_symptom)
        self.remove_symptom_btn.clicked.connect(self.remove_symptom)
        self.symptom_input.returnPressed.connect(self.add_symptom)
    
    def add_symptom(self):
        """Add a symptom with severity to the list."""
        symptom_text = self.symptom_input.text().strip()
        if symptom_text:
            severity = self.severity_input.currentText()
            list_item = QtWidgets.QListWidgetItem(f"{symptom_text} - {severity}")
            # Store both symptom and severity
            list_item.setData(QtCore.Qt.UserRole, (symptom_text, severity))
            self.symptoms_list.addItem(list_item)
            self.symptom_input.clear()
            self.symptom_input.setFocus()
    
    def remove_symptom(self):
        """Remove the selected symptom from the list."""
        selected_items = self.symptoms_list.selectedItems()
        if selected_items:
            for item in selected_items:
                self.symptoms_list.takeItem(self.symptoms_list.row(item))
    
    def save_patient(self):
        # Validate inputs
        if not self.validate_inputs():
            return
        
        # Get values from form
        patient_name = self.name_input.text().strip()
        disease_name = self.disease_input.text().strip()
        birth_date = self.birth_date_input.date().toString("yyyy-MM-dd")
        admission_date = self.admission_date_input.date().toString("yyyy-MM-dd")
        status = self.status_input.currentText()
        notes = self.notes_input.toPlainText().strip()
        
        # Get all symptoms with severity
        symptoms = []
        for i in range(self.symptoms_list.count()):
            item = self.symptoms_list.item(i)
            symptom_data = item.data(QtCore.Qt.UserRole)
            symptoms.append(symptom_data)
        
        try:
            # API endpoint URL - adjust to your server location
            api_url = "http://localhost/patient_api/add_patient.php"
            
            # Prepare payload
            payload = {
                'name': patient_name,
                'disease_name': disease_name,
                'birth_date': birth_date,
                'admission_date': admission_date,
                'status': status,
                'notes': notes,
                'symptoms': symptoms
            }
            
            # Print debug info
            print(f"Sending request to: {api_url}")
            print(f"Payload: {json.dumps(payload, indent=2)}")
            
            # Send POST request to PHP API
            response = requests.post(api_url, json=payload)
            
            # Print response for debugging
            print(f"Response status: {response.status_code}")
            print(f"Response content: {response.text}")
            
            # Check response
            if response.status_code == 200:
                result = response.json()
                if result.get('success'):
                    # Show success message
                    QtWidgets.QMessageBox.information(
                        self, "Success", 
                        f"Patient '{patient_name}' has been added successfully!"
                    )
                    # Close the dialog with Accept result
                    self.accept()
                else:
                    # Show error message from API
                    QtWidgets.QMessageBox.critical(
                        self, "API Error", 
                        f"Error: {result.get('message', 'Unknown error')}"
                    )
            else:
                # Show error message for bad response
                QtWidgets.QMessageBox.critical(
                    self, "API Error", 
                    f"Error connecting to API: Status code {response.status_code}\n{response.text}"
                )
                
        except Exception as e:
            print(f"Error saving patient: {e}")
            QtWidgets.QMessageBox.critical(
                self, "Error", 
                f"Error sending data to server: {str(e)}"
            )
    
    def validate_inputs(self):
        """Validate the form inputs."""
        # Check if patient name is provided
        if not self.name_input.text().strip():
            QtWidgets.QMessageBox.warning(
                self, "Input Error", 
                "Please enter the patient's name."
            )
            self.name_input.setFocus()
            return False
        
        # Check if disease is provided
        if not self.disease_input.text().strip():
            QtWidgets.QMessageBox.warning(
                self, "Input Error", 
                "Please enter the patient's disease."
            )
            self.disease_input.setFocus()
            return False
        
        # All validations passed
        return True


# Test code to run this module directly
if __name__ == "__main__":
    app = QtWidgets.QApplication(sys.argv)
    dialog = AddPatientDialog()
    dialog.show()
    sys.exit(app.exec_())

File: pages/addpatient.php
<?php
// Set headers for JSON response
header('Content-Type: application/json');

// Allow cross-origin requests (if needed)
header('Access-Control-Allow-Origin: *');
header('Access-Control-Allow-Methods: POST');
header('Access-Control-Allow-Headers: Content-Type');

// Handle preflight OPTIONS request
if ($_SERVER['REQUEST_METHOD'] === 'OPTIONS') {
    http_response_code(200);
    exit;
}

// Only accept POST requests
if ($_SERVER['REQUEST_METHOD'] !== 'POST') {
    http_response_code(405);
    echo json_encode(['success' => false, 'message' => 'Method not allowed']);
    exit;
}

// Get JSON input from POST request
$json_data = file_get_contents('php://input');
$data = json_decode($json_data, true);

// Debug: Log the received data
error_log("Received data: " . print_r($data, true));

// Check if data was properly decoded
if ($data === null) {
    http_response_code(400);
    echo json_encode(['success' => false, 'message' => 'Invalid JSON data']);
    exit;
}

// Validate required fields
if (empty($data['name']) || empty($data['disease_name']) || empty($data['birth_date']) || empty($data['admission_date'])) {
    http_response_code(400);
    echo json_encode(['success' => false, 'message' => 'Missing required fields']);
    exit;
}

// Connect to Oracle database using OCI8
try {
    // Oracle connection details
    // $username = 'system';
    // $password = 'Abdo2004@';
    // $connection_string = 'localhost:1521/FREE';
    $username = 'system';
    $password = 's2004b22';
    $connection_string = '192.168.21.1:1521/FREE';
    
    // Create connection
    $conn = oci_connect($username, $password, $connection_string);
    
    if (!$conn) {
        $e = oci_error();
        throw new Exception('Database connection error: ' . $e['message']);
    }
    
    // Begin transaction - we'll use OCI_DEFAULT for all executions to allow rollback
    
    // Format dates for Oracle
    $birth_date = date('d-M-Y', strtotime($data['birth_date']));
    $admission_date = date('d-M-Y', strtotime($data['admission_date']));
    
    // Insert into PATIENT table
    $patient_query = "INSERT INTO patient (name, birth_date, admission_date, status, notes) 
                     VALUES (:name, TO_DATE(:birth_date, 'DD-MON-YYYY'), 
                            TO_DATE(:admission_date, 'DD-MON-YYYY'), :status, :notes)
                     RETURNING patient_id INTO :patient_id";
    
    $patient_stmt = oci_parse($conn, $patient_query);
    
    if (!$patient_stmt) {
        $e = oci_error($conn);
        throw new Exception('Failed to prepare patient statement: ' . $e['message']);
    }
    
    // Set default status if empty
    $status = !empty($data['status']) ? $data['status'] : 'STABLE';
    $notes = !empty($data['notes']) ? $data['notes'] : '';
    
    // Bind parameters
    oci_bind_by_name($patient_stmt, ':name', $data['name']);
    oci_bind_by_name($patient_stmt, ':birth_date', $birth_date);
    oci_bind_by_name($patient_stmt, ':admission_date', $admission_date);
    oci_bind_by_name($patient_stmt, ':status', $status);
    oci_bind_by_name($patient_stmt, ':notes', $notes);
    
    // Bind the output parameter
    $patient_id = 0;
    oci_bind_by_name($patient_stmt, ':patient_id', $patient_id, -1, SQLT_INT);
    
    // Execute the patient insert
    $result = oci_execute($patient_stmt, OCI_DEFAULT);
    
    if (!$result) {
        $e = oci_error($patient_stmt);
        throw new Exception('Failed to insert patient: ' . $e['message']);
    }
    
    // Close patient statement
    oci_free_statement($patient_stmt);
    
    // Insert into DISEASE table
    $disease_query = "INSERT INTO disease (patient_id, disease_name) VALUES (:patient_id, :disease_name)";
    $disease_stmt = oci_parse($conn, $disease_query);
    
    if (!$disease_stmt) {
        $e = oci_error($conn);
        throw new Exception('Failed to prepare disease statement: ' . $e['message']);
    }
    
    // Bind parameters
    oci_bind_by_name($disease_stmt, ':patient_id', $patient_id);
    oci_bind_by_name($disease_stmt, ':disease_name', $data['disease_name']);
    
    // Execute disease insert
    $result = oci_execute($disease_stmt, OCI_DEFAULT);
    
    if (!$result) {
        $e = oci_error($disease_stmt);
        throw new Exception('Failed to insert disease: ' . $e['message']);
    }
    
    // Close disease statement
    oci_free_statement($disease_stmt);
    
    // Insert symptoms if any
    if (!empty($data['symptoms']) && is_array($data['symptoms'])) {
        $symptom_query = "INSERT INTO symptoms (patient_id, symptom_name, severity) 
                         VALUES (:patient_id, :symptom_name, :severity)";
        
        foreach ($data['symptoms'] as $symptom_data) {
            // Each symptom should be an array with name and severity
            $symptom_name = is_array($symptom_data) ? $symptom_data[0] : $symptom_data;
            $severity = is_array($symptom_data) ? $symptom_data[1] : 'Mild';
            
            $symptom_stmt = oci_parse($conn, $symptom_query);
            
            if (!$symptom_stmt) {
                $e = oci_error($conn);
                throw new Exception('Failed to prepare symptom statement: ' . $e['message']);
            }
            
            // Bind parameters
            oci_bind_by_name($symptom_stmt, ':patient_id', $patient_id);
            oci_bind_by_name($symptom_stmt, ':symptom_name', $symptom_name);
            oci_bind_by_name($symptom_stmt, ':severity', $severity);
            
            // Execute symptom insert
            $result = oci_execute($symptom_stmt, OCI_DEFAULT);
            
            if (!$result) {
                $e = oci_error($symptom_stmt);
                throw new Exception('Failed to insert symptom: ' . $e['message']);
            }
            
            // Free symptom statement
            oci_free_statement($symptom_stmt);
        }
    }
    
    // Commit transaction
    $commit = oci_commit($conn);
    
    if (!$commit) {
        $e = oci_error($conn);
        throw new Exception('Failed to commit transaction: ' . $e['message']);
    }
    
    // Close connection
    oci_close($conn);
    
    // Return success response
    http_response_code(200);
    echo json_encode([
        'success' => true, 
        'message' => 'Patient added successfully',
        'patient_id' => $patient_id
    ]);
    
} catch (Exception $e) {
    // If connection exists, rollback transaction
    if (isset($conn) && $conn) {
        oci_rollback($conn);
        oci_close($conn);
    }
    
    // Return error response with detailed information
    http_response_code(500);
    echo json_encode([
        'success' => false, 
        'message' => $e->getMessage()
    ]);
    
    // Log the error
    error_log("Patient API error: " . $e->getMessage());
}
?>